### ISSUE TITLE
we could use a windows port

### ISSUE THREAD

**Issue by presotto** [upspin/upspin#228]
Linux and MacOS aren't enough.  a volunteer would be appreciated.

**Comment by kardianos** [comment]
You'll want to decide if you want to support running it as a native windows service or if you will require some other type of superviser on windows.

If you want to go the native windows service route, these may be useful:

https://github.com/kardianos/service
https://github.com/kardianos/minwinsvc

**Comment by robpike** [comment]
Does anyone know if there is a good user-space file system option for Windows analogous to FUSE? I am embarrassed I know very little about systems programming on Windows.

**Comment by presotto** [comment]
There's Dokany but I've never used it so I can't tell you how good, bad, or
easy it is.  I'd love to hear from someone that has.

**Comment by kardianos** [comment]
@robpike I've used https://github.com/dokan-dev/dokany (well, doken 0.6) a number of years ago. It worked well for my use. Users would still need to install the dokany driver first. https://github.com/dokan-dev/dokany . This fork provides a fuse compatible C interface you can use, but I would probably recommend doing the normal import user-space-doken.dll and access methods like most windows syscalls use.

Windows will also represent WebDAV as a system drive, but before windows 10 there are a number of awkward bugs in the windows implementation that make it somewhat painful at times. That's all I'm aware of (and I've looked for various projects over the years).

Issues I've had with windows WebDAV built-in client:
 * Pre-windows 10 doesn't handle non root "/" http mount point well.
 * All implementations don't seem to like sending basic auth creds to any https endpoint without a registry entry change.
My general advice is to avoid the windows WebDAV client.

In short, the options aren't great in this area. Most "professional" products end up creating a full file system provider loaded as a DLL. Maybe we can create one in Go eventually. ;)

**Comment by ctd1500** [comment]
Keybase makes use of Dokan with Go for their encrypted file system: https://github.com/keybase/kbfs

**Comment by kwirk** [comment]
This looks interesting: https://github.com/billziss-gh/winfsp

Has unusual licensing terms: GPLv3 but exception allowing non-copyleft terms to FLOSS projects.

**Comment by n2vi** [comment]
I'm still busy on the OpenBSD port personally, but for anyone working on Windows10 port here's what I currently see:



PS C:\Users\Eric\s\src\upspin.io> go test -short ./...
ok      upspin.io/access        0.119s
ok      upspin.io/bind  0.608s
ok      upspin.io/cache 0.164s
ok      upspin.io/client        0.846s
ok      upspin.io/client/clientutil     0.204s
ok      upspin.io/client/file   0.104s
ok      upspin.io/cloud/gcpmetric       0.208s
?       upspin.io/cloud/https   [no test files]
?       upspin.io/cloud/log     [no test files]
?       upspin.io/cloud/mail    [no test files]
ok      upspin.io/cloud/mail/sendgrid   0.173s
ok      upspin.io/cloud/storage 0.177s
?       upspin.io/cloud/storage/disk    [no test files]
ok      upspin.io/cloud/storage/gcs     0.207s
?       upspin.io/cloud/storage/storagetest     [no test files]
?       upspin.io/cmd/cacheserver       [no test files]
?       upspin.io/cmd/cacheserver/cacheutil     [no test files]
?       upspin.io/cmd/dirserver [no test files]
--- FAIL: TestDocList (0.01s)
        frontend_test.go:112: expected response body to contain "<a href=\"/doc/test.md\">Test</a>"; body: "<!DOCTYPE html>\r\n\r\n<meta charset=\"utf-8\">\r\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\r\n<link rel=\"icon\" type=\"image/png\" href=\"/images/favicon-32x32.png\" sizes=\"32x32\">\r\n<title>Upspin</title>\r\n<style>\r\n*,\r\n:before,\r\n:after {\r\n  box-sizing: border-box;\r\n}\r\nhtml,\r\nbody {\r\n  height: 100%;\r\n}\r\nbody {\r\n  display: flex;\r\n  flex-direction: column;\r\n  font: 14px arial, sans-serif;\r\n  line-height: 1.45;\r\n  margin: 0;\r\n  padding: 0;\r\n}\r\nfooter {\r\n  color: #666;\r\n  font-size: .9em;\r\n  padding: 1em 1em 2em;\r\n  text-align: center;\r\n}\r\nmain {\r\n  flex: 1;\r\n  padding: 0 1em;\r\n}\r\n.container {\r\n  margin: 0 auto;\r\n  max-width: 50em;\r\n}\r\n.homeLink:link,\r\n.homeLink:visited {\r\n  color: #000;\r\n  text-decoration: none;\r\n}\r\n\r\nul {\r\n  list-style: none;\r\n  margin: 0;\r\n  padding: 0;\r\n}\r\n\r\n</style>\r\n<main>\r\n  <div class=\"container\">\r\n    \r\n<h2>Documentation</h2>\r\n<ul>\r\n<li><a href=\"/doc/test.md\">Test\r</a></li>\r\n</ul>\r\n\r\n  </div>\r\n</main>\r\n<footer>\r\n  <div>\r\n    Except as\r\n    <a href=\"https://developers.google.com/site-policies#restrictions\">noted</a>,\r\n    the content of this page is licensed under the\r\n    Creative Commons Attribution 3.0 License,\r\n    and code is licensed under a\r\n    <a href=\"https://github.com/upspin/upspin/blob/master/LICENSE\">BSD license</a>.\r\n  </div>\r\n  <div>\r\n    The Upspin mascot is Copyright 2017 Renee French.\r\n    <a href=\"https://upspin.io/doc/mascot.md\">All Rights\r\n    Reserved.</a>\r\n  </div>\r\n  <div>\r\n    <a href=\"/doc/tos.md\">Terms of Service</a> |\r\n    <a href=\"http://www.google.com/intl/en/policies/privacy/\">Privacy Policy</a>\r\n  </div>\r\n</footer>\r\n"
--- FAIL: TestDoc (0.00s)
        frontend_test.go:125: expected response body to contain "<title>Test · Upspin</title>"; body: "<!DOCTYPE html>\r\n\r\n<meta charset=\"utf-8\">\r\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\r\n<link rel=\"icon\" type=\"image/png\" href=\"/images/favicon-32x32.png\" sizes=\"32x32\">\r\n<title>Test\r · Upspin</title>\r\n<style>\r\n*,\r\n:before,\r\n:after {\r\n  box-sizing: border-box;\r\n}\r\nhtml,\r\nbody {\r\n  height: 100%;\r\n}\r\nbody {\r\n  display: flex;\r\n  flex-direction: column;\r\n  font: 14px arial, sans-serif;\r\n  line-height: 1.45;\r\n  margin: 0;\r\n  padding: 0;\r\n}\r\nfooter {\r\n  color: #666;\r\n  font-size: .9em;\r\n  padding: 1em 1em 2em;\r\n  text-align: center;\r\n}\r\nmain {\r\n  flex: 1;\r\n  padding: 0 1em;\r\n}\r\n.container {\r\n  margin: 0 auto;\r\n  max-width: 50em;\r\n}\r\n.homeLink:link,\r\n.homeLink:visited {\r\n  color: #000;\r\n  text-decoration: none;\r\n}\r\n\r\nimg {\r\n  max-width: 100%;\r\n}\r\ntable{\r\n  border-collapse: collapse;\r\n  width: 100%;\r\n}\r\ntd {\r\n  border: 1px solid #666;\r\n  padding: .25em .45em;\r\n}\r\n\r\n</style>\r\n<main>\r\n  <div class=\"container\">\r\n    <h1>Test</h1>\n\r\n  </div>\r\n</main>\r\n<footer>\r\n  <div>\r\n    Except as\r\n    <a href=\"https://developers.google.com/site-policies#restrictions\">noted</a>,\r\n    the content of this page is licensed under the\r\n    Creative Commons Attribution 3.0 License,\r\n    and code is licensed under a\r\n    <a href=\"https://github.com/upspin/upspin/blob/master/LICENSE\">BSD license</a>.\r\n  </div>\r\n  <div>\r\n    The Upspin mascot is Copyright 2017 Renee French.\r\n    <a href=\"https://upspin.io/doc/mascot.md\">All Rights\r\n    Reserved.</a>\r\n  </div>\r\n  <div>\r\n    <a href=\"/doc/tos.md\">Terms of Service</a> |\r\n    <a href=\"http://www.google.com/intl/en/policies/privacy/\">Privacy Policy</a>\r\n  </div>\r\n</footer>\r\n"
FAIL
FAIL    upspin.io/cmd/frontend  0.345s
?       upspin.io/cmd/keyserver [no test files]
?       upspin.io/cmd/storeserver       [no test files]
ok      upspin.io/cmd/upspin    0.309s
?       upspin.io/cmd/upspin-deploy     [no test files]
?       upspin.io/cmd/upspin-setupstorage       [no test files]
?       upspin.io/cmd/upspinserver      [no test files]
ok      upspin.io/config        0.180s
--- FAIL: TestLogFile (0.00s)
        log_test.go:138: creating test directory
--- FAIL: TestLogGlob (0.00s)
        log_test.go:229: creating test directory
FAIL
FAIL    upspin.io/dir/dircache  0.182s
ok      upspin.io/dir/inprocess 2.489s
?       upspin.io/dir/remote    [no test files]
2017/03/20 14:12:54.097622 Created root: fred@flintstone.org/
2017/03/20 14:12:54.105627 Deleting root "fred@flintstone.org/"
--- FAIL: TestMakeRoot (0.12s)
        all_test.go:85: fred@flintstone.org/: dir/server.Delete: I/O error:
                        dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\DirServer216330219\tree.log.fred@flintstone.org: The process cannot access the file because it is being used by another process.
2017/03/20 14:12:54.132645 Created root: wilma@flintstone.org/
--- FAIL: TestPut (0.00s)
        all_test.go:127: fred@flintstone.org/file1.txt: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestMakeDirectory (0.00s)
        all_test.go:145: fred@flintstone.org/dir: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestLink (0.00s)
        all_test.go:179: fred@flintstone.org/mylink: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestWhichAccess (0.00s)
        all_test.go:275: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestHasRight (0.00s)
        all_test.go:332: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestGlobDoesNotRemoveRoot (0.00s)
        all_test.go:371: dir/server.Glob: item does not exist:
                        fred@flintstone.org/:
                        dir/server.listDir:
                        dir/server/tree.Flush:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
2017/03/20 14:12:54.156661 Created root: joe@somewhere.com/
--- FAIL: TestGlob (0.00s)
        all_test.go:444: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestDeletePermission (0.00s)
        all_test.go:702: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestDelete (0.00s)
        all_test.go:744: err = dir/server/tree.Delete: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user, want = directory not empty
--- FAIL: TestForgetsRemoteGroupFiles (0.00s)
        all_test.go:782: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestPermissionDenied (0.01s)
        all_test.go:841: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestOverwriteFileWithWrongSequence (0.00s)
        all_test.go:882: fred@flintstone.org/Access: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
--- FAIL: TestPutBadGroup (0.00s)
        all_test.go:916: fred@flintstone.org/Group: dir/server.Put: item does not exist:
                        user fred@flintstone.org: dir/server/tree.LogIndex.Root: no root for user
2017/03/20 14:12:54.207696 Created root: bambi@forest.earth/
2017/03/20 14:12:54.232712 Created root: bambi+snapshot@forest.earth/
2017/03/20 14:12:54.277742 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/01:00"
2017/03/20 14:12:54.360798 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/14:00"
2017/03/20 14:12:54.432846 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/17:01"
2017/03/20 14:12:54.445854 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/17:02"
2017/03/20 14:12:54.591951 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/17:03"
2017/03/20 14:12:54.648991 dir/server: Snapshotted "bambi@forest.earth/" into "bambi+snapshot@forest.earth/2017/01/02/17:13"
2017/03/20 14:12:54.709030 Created root: user+snapshot@example.com/
FAIL
FAIL    upspin.io/dir/server    0.941s
ok      upspin.io/dir/server/tree       1.634s
?       upspin.io/dir/unassigned        [no test files]
ok      upspin.io/errors        2.906s
?       upspin.io/exp/client/gobind     [no test files]
?       upspin.io/exp/cmd/upbox [no test files]
?       upspin.io/exp/dir/filesystem    [no test files]
?       upspin.io/exp/store/filesystem  [no test files]
ok      upspin.io/factotum      0.238s
?       upspin.io/flags [no test files]
ok      upspin.io/key/inprocess 0.094s
ok      upspin.io/key/proquint  0.128s
?       upspin.io/key/remote    [no test files]
--- FAIL: TestLog (0.48s)
        log_test.go:122: generated log of length 520143, want 522559
2017/03/20 14:12:59.013901 key/server.Put: error fetching "uncool@buddy.com": item does not exist
2017/03/20 14:12:59.014902 key/server.Put: error fetching "uncool@buddy.com": item does not exist
2017/03/20 14:12:59.014902 key/server.Put: error fetching "cool@dude.com": item does not exist
2017/03/20 14:12:59.014902 key/server.Put: error fetching "cool+suffix@dude.com": item does not exist
2017/03/20 14:12:59.068938 key/server.Put: error fetching "other@dude.com": item does not exist
2017/03/20 14:12:59.070940 key/server.Put: error fetching "other@dude.com": item does not exist
2017/03/20 14:12:59.070940 key/server.Put: error fetching "other@dude.com": item does not exist
FAIL
FAIL    upspin.io/key/server    0.711s
ok      upspin.io/key/sha256key 0.098s
?       upspin.io/key/transports        [no test files]
?       upspin.io/key/unassigned        [no test files]
ok      upspin.io/key/usercache 0.219s
ok      upspin.io/log   0.087s
ok      upspin.io/metric        0.080s
?       upspin.io/pack  [no test files]
ok      upspin.io/pack/ee       1.827s
ok      upspin.io/pack/eeintegrity      0.672s
?       upspin.io/pack/internal [no test files]
?       upspin.io/pack/internal/packtest        [no test files]
?       upspin.io/pack/packutil [no test files]
ok      upspin.io/pack/plain    0.444s
ok      upspin.io/path  0.094s
# upspin.io/shutdown
shutdown\shutdown_test.go:112: undefined: syscall.Kill
ok      upspin.io/rpc   4.055s
?       upspin.io/rpc/dirserver [no test files]
?       upspin.io/rpc/keyserver [no test files]
?       upspin.io/rpc/storeserver       [no test files]
ok      upspin.io/serverutil    0.506s
2017/03/20 14:13:03.014570 Created root: aly@example.com/
2017/03/20 14:13:03.563937 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:04.029248 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:04.265405 serverutil/perm: Setting writers to: [aly@example.com bob@uncle.com]
2017/03/20 14:13:04.273410 Created root: bob@uncle.com/
2017/03/20 14:13:04.344459 Created root: aly@example.com/
2017/03/20 14:13:04.422509 serverutil/perm.NewFromDir: aly@example.com/Group/Writers: dir/server.Lookup: information withheld
2017/03/20 14:13:04.530582 Created root: aly@example.com/
2017/03/20 14:13:04.831783 Created root: aly@example.com/
2017/03/20 14:13:05.238053 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:05.301096 Created root: aly@example.com/
2017/03/20 14:13:05.486219 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:05.878481 serverutil/perm: Setting writers to: [*@superusers.com aly@example.com bob@uncle.com]
2017/03/20 14:13:06.007567 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:06.106633 serverutil/perm: Setting writers to: [*@superusers.com aly@example.com bob@uncle.com]
--- FAIL: TestAllowsOthersAndWildcard (0.94s)
        perm_test.go:214: bob@uncle.com is allowed; expected not allowed
        perm_test.go:214: master@superusers.com is allowed; expected not allowed
2017/03/20 14:13:06.230716 Created root: aly@example.com/
2017/03/20 14:13:06.294758 serverutil/perm.NewFromDir: aly@example.com/Group/Writers: dir/server.Lookup: information withheld
2017/03/20 14:13:06.355799 Created root: aly@example.com/
2017/03/20 14:13:06.514905 serverutil/perm.NewFromDir: aly@example.com/Group/Writers: dir/server.Lookup: information withheld
2017/03/20 14:13:07.085286 serverutil/perm: Setting writers to: [aly@example.com bob@uncle.com]
2017/03/20 14:13:07.129315 Created root: aly@example.com/
2017/03/20 14:13:07.245393 Created root: aly@example.com/
2017/03/20 14:13:07.587622 serverutil/perm: Setting writers to: [aly@example.com]
2017/03/20 14:13:07.613638 Created root: aly@example.com/
2017/03/20 14:13:08.110969 serverutil/perm: Setting writers to: [aly@example.com bob@uncle.com random@dude.io]
2017/03/20 14:13:08.143992 Created root: aly@example.com/
2017/03/20 14:13:08.373144 serverutil/perm: Setting writers to: [*@example.com]
2017/03/20 14:13:08.419176 Created root: aly@example.com/
2017/03/20 14:13:08.615307 serverutil/perm: Setting writers to: [aly@example.com]
FAIL
FAIL    upspin.io/serverutil/perm       5.995s
FAIL    upspin.io/shutdown [build failed]
?       upspin.io/store/inprocess       [no test files]
?       upspin.io/store/remote  [no test files]
ok      upspin.io/store/server  0.263s
?       upspin.io/store/storecache      [no test files]
?       upspin.io/store/transports      [no test files]
?       upspin.io/store/unassigned      [no test files]
?       upspin.io/subcmd        [no test files]
plain
eeintegrity
ee
2017/03/20 14:13:04.957867 Created root: test@upspin.io/
2017/03/20 14:13:04.980882 Created root: test-friend@upspin.io/
2017/03/20 14:13:04.991889 Created root: joe@middleman.com/
2017/03/20 14:13:05.544258 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:13.450979 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:20.138927 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:26.703085 Created root: test@upspin.io/
2017/03/20 14:13:26.782027 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:27.293457 Created root: test-friend@upspin.io/
2017/03/20 14:13:27.311132 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:27.412128 Created root: test-friend@upspin.io/
2017/03/20 14:13:27.456967 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:28.528610 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: Unmarshal buffer too short
2017/03/20 14:13:28.528610 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: Unmarshal buffer too short
2017/03/20 14:13:29.655579 Created root: test+snapshot@upspin.io/
2017/03/20 14:13:29.687644 dir/server: Snapshotted "test@upspin.io/" into "test+snapshot@upspin.io/2017/03/20/14:13"
2017/03/20 14:13:29.802748 dir/server.Delete: Error removing group file: access.RemoveGroup: item does not exist: group does not exist
2017/03/20 14:13:30.244807 Created root: test@upspin.io/
2017/03/20 14:13:30.319966 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:30.805562 Created root: test-friend@upspin.io/
2017/03/20 14:13:30.809232 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:30.908747 Created root: test-friend@upspin.io/
2017/03/20 14:13:30.937194 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:32.006119 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: Unmarshal buffer too short
2017/03/20 14:13:32.006119 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: Unmarshal buffer too short
2017/03/20 14:13:33.134017 Created root: test+snapshot@upspin.io/
2017/03/20 14:13:33.163977 dir/server: Snapshotted "test@upspin.io/" into "test+snapshot@upspin.io/2017/03/20/14:13"
2017/03/20 14:13:33.267029 dir/server.Delete: Error removing group file: access.RemoveGroup: item does not exist: group does not exist
2017/03/20 14:13:33.725721 Created root: test@upspin.io/
2017/03/20 14:13:33.810389 errors.E: unqualified type for "test@upspin.io/get-link-errors/src/link" from C:/Users/Eric/s/src/upspin.io/test/error_test.go:215
2017/03/20 14:13:34.319384 Created root: test-friend@upspin.io/
2017/03/20 14:13:34.323388 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:34.424024 Created root: test-friend@upspin.io/
2017/03/20 14:13:34.452924 Deleting root "test-friend@upspin.io/"
2017/03/20 14:13:35.523662 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: invalid entry size: -27
2017/03/20 14:13:35.523662 watch: sending error to client: dir/server/tree.sendEventFromLog: invalid operation: cannot read log at order 777: dir/server/tree.LogEntry.unmarshal: I/O error: invalid entry size: -27
2017/03/20 14:13:36.651074 Created root: test+snapshot@upspin.io/
2017/03/20 14:13:36.683835 dir/server: Snapshotted "test@upspin.io/" into "test+snapshot@upspin.io/2017/03/20/14:13"
2017/03/20 14:13:36.796215 dir/server.Delete: Error removing group file: access.RemoveGroup: item does not exist: group does not exist
--- FAIL: TestIntegration (32.21s)
    --- FAIL: TestIntegration/kind=server (10.57s)
        --- FAIL: TestIntegration/kind=server/packing=plain (3.54s)
            --- FAIL: TestIntegration/kind=server/packing=plain/RootDeletion (0.03s)
                integration_test.go:269: integration_test.go:267: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver793891512\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
            --- FAIL: TestIntegration/kind=server/packing=plain/GroupAccess (0.09s)
                access_test.go:556: access_test.go:547: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver793891512\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
                integration_test.go:373: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver793891512\tree.index.test+snapshot@upspin.io: The process cannot access the file because it is being used by another process.
        --- FAIL: TestIntegration/kind=server/packing=eeintegrity (3.48s)
            --- FAIL: TestIntegration/kind=server/packing=eeintegrity/RootDeletion (0.01s)
                integration_test.go:269: integration_test.go:267: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054619831\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
            --- FAIL: TestIntegration/kind=server/packing=eeintegrity/GroupAccess (0.07s)
                access_test.go:556: access_test.go:547: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054619831\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
                integration_test.go:373: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054619831\tree.index.test+snapshot@upspin.io: The process cannot access the file because it is being used by another process.
        --- FAIL: TestIntegration/kind=server/packing=ee (3.55s)
            --- FAIL: TestIntegration/kind=server/packing=ee/RootDeletion (0.01s)
                integration_test.go:269: integration_test.go:267: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054878378\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
            --- FAIL: TestIntegration/kind=server/packing=ee/GroupAccess (0.07s)
                access_test.go:556: access_test.go:547: client.Delete: I/O error:
                                test-friend@upspin.io/: dir/server.Delete:
                                dir/server/tree.DeleteLogs: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054878378\tree.log.test-friend@upspin.io: The process cannot access the file because it is being used by another process.
                integration_test.go:373: remove C:\Users\Eric\AppData\Local\Temp\testenv-dirserver054878378\tree.index.test+snapshot@upspin.io: The process cannot access the file because it is being used by another process.
FAIL
FAIL    upspin.io/test  32.797s
?       upspin.io/test/servermux        [no test files]
?       upspin.io/test/testenv  [no test files]
?       upspin.io/test/testfixtures     [no test files]
?       upspin.io/test/testutil [no test files]
?       upspin.io/transports    [no test files]
ok      upspin.io/upspin        0.100s
?       upspin.io/upspin/proto  [no test files]
ok      upspin.io/user  0.087s
ok      upspin.io/valid 0.118s

PS C:\Users\Eric\s\src\upspin.io> go env
set GOARCH=amd64
set GOBIN=C:\Users\Eric\s\bin
set GOEXE=.exe
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOOS=windows
set GOPATH=C:\Users\Eric\s
set GORACE=
set GOROOT=C:\Go
set GOTOOLDIR=C:\Go\pkg\tool\windows_amd64
set GCCGO=gccgo
set CC=gcc
set GOGCCFLAGS=-m64 -mthreads -fmessage-length=0
set CXX=g++
set CGO_ENABLED=1
set PKG_CONFIG=pkg-config
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2

**Comment by n2vi** [comment]
Additional things to check on:

1. Create $HOME/.ssh if needed.

2. Debug why the setup doc's example
   upspin cp hello.jpg you@gmail.com/
gives the error message
   upspin: cp: local pattern not qualified path: hello.jpg

3. Simplify install and signup for Windows people who are not used to PowerShell.

4. Most of all, make upspinfs work.

**Comment by kris-nova** [comment]
Just joined the Azure team recently, and would really like to get involved with this issue and help out. I see some unit test fixes from @AugieBot, but would like to offer some cycles to help with Azure implementation and/or the filesystem implementation. Cheers

**Comment by robpike** [comment]
Offer accepted! Thanks. How can we best get you started? It might be worth moving this discussion to the mailing list for a while.

**Comment by kris-nova** [comment]
@robpike fabulous, will start a thread on the mailing list and go from there. Cheers

**Comment by SCKelemen** [comment]
Has there been any off-band updates on this. I come from a windows development background, and I would like to contribute to this project. I'm not affiliated with Microsoft/Windows/Azure in anyway, but I have experience with both Go and Windows. I'd like to assist, but I'll probably need a little direction at first.

**Comment by n2vi** [comment]
Any experience with Dokan?   upspinfs is the primary missing component on
Windows.

**Comment by SCKelemen** [comment]
Nope, but I am tenacious and I like to learn.
